### PR TITLE
Cleanup bad unused code warnings

### LINKED ISSUE #676: Bad warnings
I tried this code: https://godbolt.org/z/nTj8qfqTx

```rust
pub struct S(u64, u64);

impl S {
    pub fn sum(&self) -> u64 {
        self.0 + self.1
    }
}

pub fn f() -> u64 {
    S(123, 42).sum()
}

```

I expected to see this happen: no warnings produced

Instead, this happened:

```
<source>:1:5: warning: struct is never constructed: 'S'
    1 | pub struct S(u64, u64);
      |     ^
<source>:4:9: warning: associated function is never used: 'sum'
    4 |     pub fn sum(&self) -> u64 {
      |         ^
<source>:9:5: warning: function is never used: 'f'
    9 | pub fn f() -> u64 {
      |     ^
<source>:9:5: warning: unused name 'f'
Compiler returned: 0
```

### Meta

- What version of Rust GCC were you using, git sha if possible. c644ee4c4351e3590f5396e94ec24ad7b828a954

### PR DESCRIPTION
**Pull request by philberty**
This patchset contains 4 distinct fixes:

When a constant is declared after where it is used the code-generation pass falls
back to a query compilation of the HIR::Item this did not contain a check to verify
if it was already compiled and results in duplicate CONST_DECLS being generated
if query compilation was used.

We were using a zero precision integer to contain unit-type expressions this results
in VAR_DECLS being lost in the GENERIC graph which does not allow us to perform
any static analysis upon the DECL. This changes the unit type to use an empty struct
and for initialization of a VAR_DECL we can simply pass an empty constructor and let 
GCC optimize this code for us.

Update our DEAD_CODE scan to take into account modules of items and also respect
if structures are prefixed with an underscore we can ignore generating an unused warning.

Remove our AST scan for unused code and reuse GCC TREE_USED to track wether
VAR_DECL, PARM_DECL, CONST_DECL are actually used or not. We reuse the GCC
walk_tree functions to have this as nice separate lint.

Fixes #676

### CODE REVIEW

File: gcc/testsuite/rust/compile/torture/all_doc_comment_line_blocks_crlf.rs [hunk 1]
```diff
@@ -9,40 +9,37 @@
...
-    /*!*/
-    //
-
-    ///
-    mod doc { }    // { dg-warning "unused name" }
-
-    /**/
-    /***/
-  }
+pub mod module {
```
Comment (dkm) [inline gcc/testsuite/rust/compile/torture/all_doc_comment_line_blocks_crlf.rs:12]: minor comment, but looks like you are reindenting automatically here.

**Review by CohenArthur — APPROVED** [review]
Nice, this is clean. I guess we'll need to think about the default behavior/warnings we want to enable regarding unused variables, either on gccrs or over on cargo-gccrs

File: gcc/testsuite/rust/compile/torture/all_doc_comment_line_blocks_crlf.rs (hunk 1 above)
Comment (philberty) [reply gcc/testsuite/rust/compile/torture/all_doc_comment_line_blocks_crlf.rs]: good find, those edits negated the test case. I have reverted my changes and simply removed the dg-warnings properly.

File: gcc/rust/lang.opt [hunk 2]
```diff
@@ -38,6 +38,18 @@ Wall
...
+Rust Var(warn_unused_variable) Init(1) Warning
+; documented in common.opt
+
+Wunused-const-variable
+Rust Warning Alias(Wunused-const-variable=, 2, 0)
+Warn when a const variable is unused.
+
+Wunused-const-variable=
+Rust Joined RejectNegative UInteger Var(warn_unused_const_variable) Init(1) Warning LangEnabledBy(Rust,Wunused-variable, 1, 0) IntegerRange(0, 2)
+Warn when a const variable is unused.
```
Comment (tschwinge) [inline gcc/rust/lang.opt:51]: I'm back to working on resolving <https://gcc.gnu.org/PR105914> "gccrs setting warn_unused_variable breaks thousands of non-Rust tests".  I do understand what the problem is (details to be shared later), and how to resolve it.  And, I do have a set of patches for GCC options machinery, to be upstreamed separately, to make it less likely that the same issue appears again.

I do understand that you'd like `-Wunused-variable`, `-Wunused-const-variable=1` enabled by default.  (As implemented here, via `Init(1)`.)  Now I need to know: (1) should that be enforced unconditionally (that is, no way to disable those diagnostics), or (2) should it be possible to specify `-Wno-unused-variable` etc. to disable the diagnostics?  (Natural for GCC would be the latter, (2).)

**Pull request merged**